### PR TITLE
feat: add Recursive tree traversal techniques

### PR DESCRIPTION
**Pull request by Lazeeez**
#### Description of Change

All Tree traversal techniques (inorder, postorder, preorder) with recursive functions.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Added tree traversal implementations with recursive functions.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1856"></a>

### CODE REVIEW

**Comment by Lazeeez** [comment]
@Panquesito7 can you please review this PR, thanks.

File: others/recursive_tree_traversal.cpp [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+ *
+******************************************************************************/
+
+#include <iostream>   /// for I/O operations
+#include <cassert>    /// for assert
+#include <vector>     /// for vector
+
+/******************************************************************************
+ * @namespace search
+ * @brief Searching algorithms
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:60]:
```suggestion
 * @namespace others
 * @brief Other algorithms
```

File: others/recursive_tree_traversal.cpp [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+ *******************************************************************************/
+namespace others {
+
+    /******************************************************************************
+    * @namespace interpolation_search
+    * @brief Functions for the revursive tree traversal techniques.
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:66]:
```suggestion
    * @brief Functions for the Recursive version of Inorder, Preorder, and Postorder [Traversal of the Tree](https://en.wikipedia.org/wiki/Tree_traversal) algorithm implementation
```

File: others/recursive_tree_traversal.cpp [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+            return postorder_result;
+        }
+
+    }  // namespace recursive_tree_traversals
+
+}  // namespace others
+
+/*******************************************************************************
+ * @brief Self-test implementation #1
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:177]:
```suggestion
 * @brief 1st test-case
```

File: others/recursive_tree_traversal.cpp [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+ * @returns void
+ *******************************************************************************/
+static void test1(){
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:180]:
```suggestion
void test1(){
```

File: others/recursive_tree_traversal.cpp [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    for (auto i = 0; i < size; ++i) {
+        assert(actual_result_postorder[i] == result_postorder[i]);
+    }
+    std::cout << "Passed!" << std::endl;
+
+    std::cout << std::endl;
+}
+
+/*******************************************************************************
+ * @brief Self-test implementation #2
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:232]:
```suggestion
 * @brief 2nd test-case
```

File: others/recursive_tree_traversal.cpp [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+ * @returns void
+ *******************************************************************************/
+static void test2(){
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:235]:
```suggestion
void test2(){
```

File: others/recursive_tree_traversal.cpp [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+    for (auto i = 0; i < size; ++i) {
+        assert(actual_result_postorder[i] == result_postorder[i]);
+    }
+    std::cout << "Passed!" << std::endl;
+
+    std::cout << std::endl;
+}
+
+/*******************************************************************************
+ * @brief Self-test implementation #3
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:286]:
```suggestion
 * @brief 3rd test-case
```

File: others/recursive_tree_traversal.cpp [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+ * @returns void
+ *******************************************************************************/
+static void test3(){
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:289]:
```suggestion
void test3(){
```

File: others/recursive_tree_traversal.cpp [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    result_postorder = obj3.postorder(root);
+    std::cout << "Testcase #3: Postorder Traversal...";
+    for (auto i = 0; i < size; ++i) {
+        assert(actual_result_postorder[i] == result_postorder[i]);
+    }
+    std::cout << "Passed!" << std::endl;
+
+    std::cout << std::endl;
+}
+
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:335]:
```suggestion

/**
 * @brief Self-test implementations
 * @returns void
 */
static void tests() {
    std::cout << "1st test-case" << std::endl;
    test1();  // run 1st test-case
    std::cout << "2nd test-case" << std::endl;
    test2();  // run 2nd test-case
    std::cout << "3rd test-case" << std::endl;
    test3();  // run 3rd test-case
}
```

File: others/recursive_tree_traversal.cpp [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+    std::cout << "TestCase #1" << std::endl;
+    test1();                                    // run self-test implementations #1
+
+    //Testcase #2
+    std::cout << "TestCase #2" << std::endl;
+    test2();                                    // run self-test implementations #2
+
+    //Testcase #3
+    std::cout << "TestCase #3" << std::endl;
+    test3();                                    // run self-test implementations #3
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:353]:
```suggestion
    tests();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍

File: others/recursive_tree_traversal.cpp [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+ *
+ * (N) Process n itself.
+ * (L) Recursively traverse its left subtree. When this step is finished,
+ * we are back at n again.
+ * (R) Recursively traverse its right subtree. When this step is finished,
+ * we are back at n again.
+ *
+ * In normal preorder traversal, visit the left subtree before the right
+ *subtree. If we visit the right subtree before visiting the left subtree, it is
+ *referred to as reverse preorder traversal.
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:34]:
```suggestion
 * subtree. If we visit the right subtree before visiting the left subtree, it is
 * referred to as reverse preorder traversal.
```

File: others/recursive_tree_traversal.cpp [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+ *
+ * (L) Recursively traverse its left subtree. When this step is finished,
+ * we are back at n again.
+ * (N) Process n itself.
+ * (R) Recursively traverse its right subtree. When this step is finished,
+ * we are back at n again.
+ *
+ * In normal inorder traversal, we visit the left subtree before the right
+ *subtree. If we visit the right subtree before visiting the left subtree, it is
+ *referred to as reverse inorder traversal.
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:20]:
```suggestion
 * subtree. If we visit the right subtree before visiting the left subtree, it is
 * referred to as reverse inorder traversal.
```

File: others/recursive_tree_traversal.cpp [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+ *
+ * (L) Recursively traverse its left subtree. When this step is finished,
+ * we are back at n again.
+ * (R) Recursively traverse its right subtree. When this step is finished,
+ * we are back at n again.
+ * (N) Process n itself.
+ *
+ * In normal postorder traversal, visit the left subtree before the right
+ *subtree. If we visit the right subtree before visiting the left subtree, it is
+ *referred to as reverse postorder traversal.
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:48]:
```suggestion
 * subtree. If we visit the right subtree before visiting the left subtree, it is
 * referred to as reverse postorder traversal.
```

File: others/recursive_tree_traversal.cpp [hunk 14]
```diff
@@ -0,0 +1,372 @@
+/******************************************************************************
+ * @file
+ * @brief Recursive version of Inorder, Preorder, and Postorder [Traversal of
+ * the Tree] (https://en.wikipedia.org/wiki/Tree_traversal)
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:4]:
```suggestion
 * the Tree](https://en.wikipedia.org/wiki/Tree_traversal)
```

File: others/recursive_tree_traversal.cpp [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+
+/******************************************************************************
+ * @namespace others
+ * @brief Other algorithms
+ *******************************************************************************/
+namespace others {
+
+/******************************************************************************
+ * @namespace interpolation_search
+ * @brief Functions for the revursive tree traversal techniques.
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:66]:
```suggestion
 * @brief Functions for the [Recursive Traversal of
 * the Tree](https://en.wikipedia.org/wiki/Tree_traversal) implementation
```

File: others/recursive_tree_traversal.cpp [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+        root->data);  // Display the data part of the root (or current node)
+    inorder(root->right);  // Traverse the right subtree
+
+    return inorder_result;
+}
+
+/******************************************************************************
+ * @brief preorder function that will perform the preorder traversal
+ * recursively, and return the resultant vector that contain the preorder
+ *traversal of a tree.
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:144]:
```suggestion
 * traversal of a tree.
```

File: others/recursive_tree_traversal.cpp [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+    Node *node = new Node();
+    node->data = data;
+    node->left = node->right = nullptr;
+    return node;
+}
+
+/******************************************************************************
+ * @brief inorder() function that will perform the inorder traversal
+ * recursively, and return the resultant vector that contain the inorder
+ *traversal of a tree.
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:124]:
```suggestion
 * traversal of a tree.
```

File: others/recursive_tree_traversal.cpp [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+ * @param right follow up right subtree.
+ *******************************************************************************/
+struct Node {
+    uint64_t data = 0;     ///< The value/key of the node.
+    struct Node *left{};   ///< struct pointer to left subtree.
+    struct Node *right{};  ///< struct pointer to right subtree.
+};
+/******************************************************************************
+ * @brief BT used to make the entire structure of the binary tree and the
+ *functions associated with the binary tree
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:83]:
```suggestion
 * functions associated with the binary tree
```

File: others/recursive_tree_traversal.cpp [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+    preorder(root->left);   // Traverse the left subtree
+    preorder(root->right);  // Traverse the right subtree
+
+    return preorder_result;
+}
+
+/******************************************************************************
+ * @brief postorder function that will perform the postorder traversal
+ * recursively, and return the result vector that contain the postorder
+ *traversal of a tree.
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:164]:
```suggestion
 * traversal of a tree.
```

**Review by Panquesito7 — COMMENTED** [review]
Hey there! I saw some of my suggestions/comments are not implemented yet. Would you like to explain why you didn't accept them and what are your thoughts on this? Thank you! 😃

**Comment by Lazeeez** [comment]
> Hey there! I saw some of my suggestions/comments are not implemented yet. Would you like to explain why you didn't accept them and what are your thoughts on this? Thank you! 😃
> 

I'm sorry, I was a bit busy with the Christmas, committed then just now, please check, thanks.

**Comment by Panquesito7** [comment]
> > Hey there! I saw some of my suggestions/comments are not implemented yet. Would you like to explain why you didn't accept them and what are your thoughts on this? Thank you! 😃
> 
> I'm sorry, I was a bit busy with the Christmas, committed then just now, please check, thanks.

No worries! I hope you had great holidays with your family! :D

**Review by Qannaf — DISMISSED** [review]

**Comment by Lazeeez** [comment]
@Panquesito7 can you check what's the conflict with the branch, thanks.

**Comment by Panquesito7** [comment]
> @Panquesito7 can you check what's the conflict with the branch, thanks.

The pull bot tried updating your branch to the latest version which caused conflicts and as well removed all the changes you made 😅 that's why it's always best to use another branch for that. You can do that. :)

**Comment by Lazeeez** [comment]
> > @Panquesito7 can you check what's the conflict with the branch, thanks.
> 
> The pull bot tried updating your branch to the latest version which caused conflicts and as well removed all the changes you made 😅 that's why it's always best to use another branch for that. You can do that. :)

oh 😅

**Comment by Lazeeez** [comment]
I'll create a new PR with a different branch.

**Pull request closed without merging**